src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,128 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elementData;
+    private int size = 0;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        growIfNeeded();
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index != size) {
+            rangeCheck(index);
+        }
+        growIfNeeded();
+        System.arraycopy(this.elementData, index, this.elementData, index + 1, size - index);
+        this.elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (list.size() == 0) {
+            return;
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        rangeCheck(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        rangeCheck(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        rangeCheck(index);
+        T valueToReturn = get(index);
+        fastRemove(elementData, index);
+        return valueToReturn;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        final Object[] es = elementData;
+        final int size = this.size;
+        int i = 0;
+        found: {
+            if (element == null) {
+                for (; i < size; i++) {
+                    if (es[i] == null) {
+                        break found;
+                    }
+                }
+            } else {
+                for (; i < size; i++) {
+                    if (element.equals(es[i])) {
+                        break found;
+                    }
+                }
+            }
+            throw new NoSuchElementException("Element you want to remove does not exit");
```

Comment:
[using labels is a vary bad practice. you can do the same using just one loop and if inside this loop.](https://stackoverflow.com/questions/33689466/labels-in-java-bad-practice)